roslibrust/src/ros1/publisher.rs

```diff
@@ -0,0 +1,166 @@
+use super::tcpros::ConnectionHeader;
+use abort_on_drop::ChildTask;
+use roslibrust_codegen::RosMessageType;
+use std::{
+    marker::PhantomData,
+    net::{Ipv4Addr, SocketAddr},
+    sync::Arc,
+};
+use tokio::{
+    io::{AsyncReadExt, AsyncWriteExt},
+    sync::{mpsc, RwLock},
+};
+
+pub struct Publisher<T> {
+    topic_name: String,
+    sender: mpsc::Sender<Vec<u8>>,
+    phantom: PhantomData<T>,
+}
+
+impl<T: RosMessageType> Publisher<T> {
+    pub(crate) fn new(topic_name: &str, sender: mpsc::Sender<Vec<u8>>) -> Self {
+        Self {
+            topic_name: topic_name.to_owned(),
+            sender,
+            phantom: PhantomData,
+        }
+    }
+
+    pub async fn publish(&self, data: &T) -> Result<(), Box<dyn std::error::Error>> {
+        let data = serde_rosmsg::to_vec(&data)?;
```

Comment:
Holy shit... That just works. I expected there to be way more finagling there with some more complicated trait than `Serialize` being required.